### PR TITLE
Feature: Mass award badge

### PR DESCRIPTION
**Pull request by romanrizzi**
The mass award will happen in batches of 200 users.

No badge selected:

Screen Shot 2020-01-09 at 16 00 29

When a badge is selected:

Screen Shot 2020-01-09 at 16 00 48

### CODE REVIEW

File: app/assets/javascripts/admin/controllers/admin-badges-award.js.es6 [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+          data: new FormData()
+        };
+
+        options.data.append("file", file);
+
+        this.set("saving", true);
+
+        ajax(`/admin/badges/award/${this.model.id}`, options)
+          .then(() => {
+            bootbox.alert(I18n.t("admin.badges.mass_award.success"));
```
Comment (eviltrout) [inline app/assets/javascripts/admin/controllers/admin-badges-award.js.es6]: Maybe this should redirect to the front end view so they can see who was awarded the badge?

File: app/assets/javascripts/admin/controllers/admin-badges.js.es6 [hunk 2]
```diff
@@ -1,2 +1,19 @@
 import Controller from "@ember/controller";
-export default Controller.extend();
+import { inject as service } from "@ember/service";
+
+export default Controller.extend({
+  selectedRoute: "adminBadges.show",
+  router: service(),
+
+  actions: {
+    showBadge() {
```
Comment (eviltrout) [inline app/assets/javascripts/admin/controllers/admin-badges.js.es6]: Could we convert these actions to links? Buttons can have `href`s

File: app/assets/javascripts/admin/templates/badges-award.hbs [hunk 3]
```diff
@@ -0,0 +1,22 @@
+{{#d-section class="award-badge"}}
+  <form class="form-horizontal">
+    <h1>{{i18n 'admin.badges.mass_award.title'}}</h1>
+    <div class='badge-preview'>
+      {{#if model}}
+        {{icon-or-image model}}
```
Comment (eviltrout) [inline app/assets/javascripts/admin/templates/badges-award.hbs]: I noticed in the screenshot that this is extremely big. Could we shrink it a bit?

File: app/controllers/admin/badges_controller.rb [hunk 4]
```diff
@@ -33,6 +35,38 @@ def new
...
+
+    batch_number = 1
+    batch = []
+
+    File.open(csv_file) do |csv|
+      csv.each_line do |email_line|
+        batch.concat CSV.parse_line(email_line)
+
+        # Split the emails in batches of 200 elements.
+        full_batch = csv.lineno % (BadgeGranter::MAX_ITEMS_FOR_DELTA * batch_number) == 0
```
Comment (eviltrout) [inline app/controllers/admin/badges_controller.rb]: I wonder if 200 emails will be too large as a redis value for sidekiq? Might be ok, but we should definitely test with batches larger than 200 and perhaps tune it.

File: app/services/badge_granter.rb [hunk 5]
```diff
@@ -12,6 +12,40 @@ def self.grant(badge, user, opts = {})
...
   end
 
+  def self.mass_grant(badge, users)
+    return unless badge.enabled?
+
+    system_user_id = Discourse.system_user.id
+    user_badges = users.map { |u| { badge_id: badge.id, user_id: u.id, granted_by_id: system_user_id, granted_at: Time.now } }
+    granted_badges = UserBadge.insert_all(user_badges, returning: %i[user_id])
+
+    users_to_notify = users.each do |user|
```
Comment (eviltrout) [inline app/services/badge_granter.rb]: This variable seems unused?

File: app/services/badge_granter.rb [hunk 6]
```diff
@@ -12,6 +12,40 @@ def self.grant(badge, user, opts = {})
...
+      use_default_locale = !SiteSetting.allow_user_locale && user.locale.blank?
+      notification_locale = use_default_locale ? SiteSetting.default_locale : user.locale
+
+      notification = I18n.with_locale(notification_locale) do
```
Comment (eviltrout) [inline app/services/badge_granter.rb]: This feels a little repetitive from the #grant method below - could you split it out to a function and use it in both places?

File: app/assets/javascripts/admin/controllers/admin-badges-award.js.es6 (hunk 1 above)
Comment (romanrizzi) [reply app/assets/javascripts/admin/controllers/admin-badges-award.js.es6]: The mass award will happen asynchronously, does the redirect still make sense?

File: app/assets/javascripts/admin/controllers/admin-badges.js.es6 (hunk 2 above)
Comment (romanrizzi) [reply app/assets/javascripts/admin/controllers/admin-badges.js.es6]: I replaced the transition with an "href". However, I still need the action to dynamically set the badges link-to url.

**Pull request merged**